### PR TITLE
PRT-730 node error handling

### PR DESCRIPTION
**Pull request by ranlavanet**

### CODE REVIEW

File: protocol/chainlib/grpc.go [hunk 1]
```diff
@@ -427,11 +442,19 @@ func (cp *GrpcChainProxy) SendNodeMsg(ctx context.Context, ch chan interface{},
 	err = conn.Invoke(connectCtx, "/"+nodeMessage.Path, msg, response, grpc.Header(&respHeaders))
 
 	if err != nil {
-		if connectCtx.Err() == context.DeadlineExceeded {
-			// Not an rpc error, return provider error without disclosing the endpoint address
-			return nil, "", nil, utils.LavaFormatError("Provider Failed Sending Message", context.DeadlineExceeded)
+		if parsedError := cp.BaseChainProxy.errorHandler.HandleNodeError(ctx, err); parsedError != nil {
```
Comment (omerlavanet) [inline protocol/chainlib/grpc.go:445]: why do we need cp.BaseChainProxy.errorHandler.HandleNodeError and not cp.HandleNodeError

File: protocol/chainlib/grpc.go [hunk 2]
```diff
@@ -384,11 +399,11 @@ func (cp *GrpcChainProxy) SendNodeMsg(ctx context.Context, ch chan interface{},
 			msgLocal := msgFactory.NewMessage(methodDescriptor.GetInputType())
 			err = proto.Unmarshal(nodeMessage.Msg, msgLocal)
 			if err != nil {
-				return nil, "", nil, err
+				return nil, "", nil, utils.LavaFormatError("Failed to unmarshal proto.Unmarshal(nodeMessage.Msg, msgLocal)", err)
```
Comment (omerlavanet) [inline protocol/chainlib/grpc.go:402]: will this spam errors even if the send is a warning like in chain fetcher fails?

File: protocol/chainlib/grpc.go [hunk 3]
```diff
@@ -427,11 +442,19 @@ func (cp *GrpcChainProxy) SendNodeMsg(ctx context.Context, ch chan interface{},
...
 	if err != nil {
-		if connectCtx.Err() == context.DeadlineExceeded {
-			// Not an rpc error, return provider error without disclosing the endpoint address
-			return nil, "", nil, utils.LavaFormatError("Provider Failed Sending Message", context.DeadlineExceeded)
+		if parsedError := cp.BaseChainProxy.errorHandler.HandleNodeError(ctx, err); parsedError != nil {
+			return nil, "", nil, parsedError
 		}
-		return nil, "", nil, utils.LavaFormatError("Invoke Failed", err, utils.Attribute{Key: "GUID", Value: ctx}, utils.Attribute{Key: "Method", Value: nodeMessage.Path}, utils.Attribute{Key: "msg", Value: nodeMessage.Msg})
+		// return the node's error back to the client as the error type is a invalid request which is cu deductible
+		respBytes, handlingError := handleGrpcNodeError(ctx, err)
```
Comment (omerlavanet) [inline protocol/chainlib/grpc.go:449]: naming is confusing

File: protocol/chainlib/grpc.go [hunk 4]
```diff
@@ -447,6 +470,23 @@ func (cp *GrpcChainProxy) SendNodeMsg(ctx context.Context, ch chan interface{},
 	return reply, "", nil, nil
 }
 
+func handleGrpcNodeError(ctx context.Context, err error) ([]byte, error) {
+	var respBytes []byte
+	var marshalingError error
+	if statusError, ok := status.FromError(err); ok {
+		respBytes, marshalingError = json.Marshal(&GrpcNodeErrorResponse{ErrorMessage: statusError.Message(), ErrorCode: uint32(statusError.Code())})
+		if marshalingError != nil {
+			return nil, utils.LavaFormatError("json.Marshal(&GrpcNodeErrorResponse Failed 1", err, utils.Attribute{Key: "GUID", Value: ctx})
```
Comment (omerlavanet) [inline protocol/chainlib/grpc.go:479]: this will spam errors

File: protocol/chainlib/jsonRPC.go [hunk 5]
```diff
@@ -435,9 +435,10 @@ func (cp *JrpcChainProxy) SendNodeMsg(ctx context.Context, ch chan interface{},
 		connectCtx, cancel := cp.NodeUrl.LowerContextTimeout(ctx, relayTimeout)
 		defer cancel()
 		rpcMessage, err = rpc.CallContext(connectCtx, nodeMessage.ID, nodeMessage.Method, nodeMessage.Params)
-		if err != nil && connectCtx.Err() == context.DeadlineExceeded {
-			// Not an rpc error, return provider error without disclosing the endpoint address
-			return nil, "", nil, utils.LavaFormatError("Provider Failed Sending Message", context.DeadlineExceeded)
+		if err != nil {
+			if parsedError := cp.BaseChainProxy.errorHandler.HandleNodeError(ctx, err); parsedError != nil {
```
Comment (omerlavanet) [inline protocol/chainlib/jsonRPC.go:439]: add comments, change HandleNodeError to say what errors are processed here and what is processed later

File: protocol/chainlib/rest.go [hunk 6]
```diff
@@ -416,9 +416,8 @@ func (rcp *RestChainProxy) SendNodeMsg(ctx context.Context, ch chan interface{},
 	}
 	res, err := httpClient.Do(req)
 	if err != nil {
-		if err != nil && connectCtx.Err() == context.DeadlineExceeded {
-			// Not an rpc error, return provider error without disclosing the endpoint address
-			return nil, "", nil, utils.LavaFormatError("Provider Failed Sending Message", context.DeadlineExceeded)
+		if parsedError := rcp.BaseChainProxy.errorHandler.HandleNodeError(ctx, err); parsedError != nil {
```
Comment (omerlavanet) [inline protocol/chainlib/rest.go:419]: can remove the .Base.errorHandler

File: protocol/chainlib/common.go [hunk 7]
```diff
@@ -181,6 +181,7 @@ type CollectionKey struct {
 }
 
 type BaseChainProxy struct {
+	errorHandler     ErrorHandler
```
Comment (omerlavanet) [inline protocol/chainlib/common.go:184]: we can embed ErrorHandler here then we don't need to access it as a member

File: protocol/chainlib/tendermintRPC.go [hunk 8]
```diff
@@ -417,7 +417,7 @@ func NewtendermintRpcChainProxy(ctx context.Context, nConns uint, rpcProviderEnd
 	}
 	websocketUrl, httpUrl := verifyTendermintEndpoint(rpcProviderEndpoint.NodeUrls)
 	cp := &tendermintRpcChainProxy{
-		JrpcChainProxy: JrpcChainProxy{BaseChainProxy: BaseChainProxy{averageBlockTime: averageBlockTime, NodeUrl: websocketUrl}, conn: map[string]*chainproxy.Connector{}},
+		JrpcChainProxy: JrpcChainProxy{BaseChainProxy: BaseChainProxy{averageBlockTime: averageBlockTime, NodeUrl: websocketUrl, errorHandler: &TendermintRPCErrorHandler{}}, conn: map[string]*chainproxy.Connector{}},
```
Comment (omerlavanet) [inline protocol/chainlib/tendermintRPC.go:420]: same comment

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: protocol/chainlib/grpc.go (hunk 2 above)
Comment (ranlavanet) [reply protocol/chainlib/grpc.go:402]: we discussed it and agreed to keep it

File: protocol/chainlib/grpc.go (hunk 4 above)
Comment (ranlavanet) [reply protocol/chainlib/grpc.go:479]: we discussed it and agreed to keep it

**Pull request merged**